internal/lang/eval.go

```diff
@@ -196,6 +199,60 @@ func (s *Scope) EvalExpr(expr hcl.Expression, wantType cty.Type) (cty.Value, tfd
 	return val, diags
 }
 
+// Identify and enhance any function related dialogs produced by a hcl.EvalContext
+func (s *Scope) enhanceFunctionDiags(diags hcl.Diagnostics) hcl.Diagnostics {
+	out := make(hcl.Diagnostics, len(diags))
+	for i, diag := range diags {
+		out[i] = diag
+
+		if funcExtra, ok := diag.Extra.(hclsyntax.FunctionCallUnknownDiagExtra); ok {
+			funcName := funcExtra.CalledFunctionName()
+			// prefix::stuff::
+			fullNamespace := funcExtra.CalledFunctionNamespace()
+
+			if !strings.Contains(fullNamespace, "::") {
+				// Not a namespaced function, no enhancements nessesary
+				continue
+			}
+
+			// Insert the an enhanced copy of diag into diags
+			enhanced := *diag
+			out[i] = &enhanced
+
+			// Update enhanced with additional details
+
+			if fullNamespace == "core::" {
+				// Error is in core namespace, mirror non-core equivalent
+				enhanced.Summary = "Call to unknown function"
+				enhanced.Detail = fmt.Sprintf("There is no builtin (core::) function named %q.", funcName)
+				continue
+			}
+
+			providerFuncNamespace := regexp.MustCompile("^([^:]*)::([^:]*)::$")
```

Comment:
Let's move the regex complilation out of the loop here for performance.

Reply:
maybe this "is a provider function namespace" kind of logic could be pulled out to it's own method?

Reply:
Moved the regex outside the function.  It's only ever needed here, probably not worth refactoring unless we have multiple use cases.